core/shared/src/main/scala/org/typelevel/log4cats/Log.scala

```diff
@@ -45,112 +45,96 @@ object Log {
   trait Builder[Ctx] {
     def withTimestamp(value: FiniteDuration): Builder[Ctx]
     def withLevel(level: KernelLogLevel): Builder[Ctx]
-    def withLevelValue(levelValue: Int): Builder[Ctx]
     def withMessage(message: => String): Builder[Ctx]
     def withThrowable(throwable: Throwable): Builder[Ctx]
-    def withContext(name: String)(f: Ctx): Builder[Ctx]
+    def withContext[A](name: String)(ctx: A)(implicit E: Context.Encoder[A, Ctx]): Builder[Ctx]
     def withFileName(name: String): Builder[Ctx]
     def withClassName(name: String): Builder[Ctx]
+    def withMethodName(name: String): Builder[Ctx]
     def withLine(line: Int): Builder[Ctx]
 
     final def withContextMap[A](
         contextMap: Map[String, A]
-    )(implicit ev: A =:= Ctx): Builder[Ctx] = {
-      var builder = this
-      contextMap.foreach { case (k, v) =>
-        builder = withContext(k)(ev(v))
-      }
-      builder
-    }
+    )(implicit E: Context.Encoder[A, Ctx]): Builder[Ctx] =
+      contextMap.foldLeft(this) { case (builder, (k, v)) => builder.withContext(k)(v) }
 
     def build(): Log[Ctx]
   }
 
   def mutableBuilder[Ctx](): Builder[Ctx] = new MutableBuilder[Ctx]()
 
-  private class MutableBuilder[Ctx] private[Log] () extends Builder[Ctx] with Log[Ctx] {
+  private val noopMessage: () => String = () => ""
+
+  private class MutableBuilder[Ctx] extends Builder[Ctx] {
     private var _timestamp: Option[FiniteDuration] = None
     private var _level: Option[KernelLogLevel] = None
-    private var _levelValue: Option[Int] = None
-    private var _message: Option[String] = None
+    private var _message: () => String = noopMessage
```

Comment:
Question: is the use case of a log without a message something we want to support?

Reply:
@kubukoz can you think of an reason we'd want to support a log with a message?

If not, this could probably be moved to a constructor parameter, and we could drop `withMessage`